src/mgm.rs

```diff
@@ -152,29 +173,63 @@ impl MgmAlgorithmId {
 ///
 /// The only supported algorithm for MGM keys is 3DES.
 #[derive(Clone)]
-pub struct MgmKey([u8; DES_LEN_3DES]);
+pub struct MgmKey(MgmKeyKind);
+
+/// Supported kinds of MGM ciphers
+#[derive(Clone)]
+enum MgmKeyKind {
+    Tdes(Key<TdesEde3>),
+}
 
 impl MgmKey {
-    /// Generate a random MGM key
-    pub fn generate() -> Self {
-        let mut key_bytes = [0u8; DES_LEN_3DES];
-        let mut rng = OsRng.unwrap_err();
-        rng.fill_bytes(&mut key_bytes);
-        Self(key_bytes)
+    /// Generates a random MGM key for the given algorithm.
+    pub fn generate(alg: MgmAlgorithmId, rng: &mut impl TryCryptoRng) -> Result<Self> {
+        match alg {
+            MgmAlgorithmId::ThreeDes => {
+                let key = TdesEde3::try_generate_key_with_rng(rng).map_err(|e| {
+                    error!("RNG failure: {}", e);
+                    Error::KeyError
+                })?;
+
+                Ok(Self(MgmKeyKind::Tdes(key)))
+            }
+        }
+    }
+
+    /// Generates a random MGM key using the preferred algorithm for the given YubiKey's
+    /// firmware version.
+    pub fn generate_for(yubikey: &YubiKey, rng: &mut impl TryCryptoRng) -> Result<Self> {
+        let alg = MgmAlgorithmId::default_for(yubikey)?;
+        Self::generate(alg, rng)
     }
 
-    /// Create an MGM key from byte slice.
+    /// Parses an MGM key from the given byte slice.
     ///
-    /// Returns an error if the slice is the wrong size or the key is weak.
-    pub fn from_bytes(bytes: impl AsRef<[u8]>) -> Result<Self> {
-        bytes.as_ref().try_into()
+    /// Returns an error if the slice is an invalid size or the key is weak.
+    ///
+    /// If `alg` is `None`, the algorithm will be selected based on the length of the
+    /// slice, returning an error if there is not a unique match.
+    pub fn from_bytes(bytes: impl AsRef<[u8]>, alg: Option<MgmAlgorithmId>) -> Result<Self> {
```

Comment:
I was wondering if instead of the `Option<MgmAlgorithmId>` approach, it should perhaps just be two methods, one which infers the algorithm from the length, and another one which always takes an `MgmAlgorithmId`.